### PR TITLE
BFS problems

### PR DESCRIPTION
**Pull request by Eli-BH**
BFS problems

### CODE REVIEW

File: eliHenderson/assignments/BreadthFirstSearch/lc102.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ *  insert the root into the queue by unshifting
+ *  while queue is not empty
+ *  capture the lvl size
+ *  create an array that holds the current levels nodes
+ *  loop through the queue lvl amounts of times
+ *  save the shifted queue node
+ *   push the node in to the temp list
+ *    check for children
+ *    unshift the children if they exist
+ *  unshift the current level into the results list
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc102.js:17]: I think you mean push here.

File: eliHenderson/assignments/BreadthFirstSearch/lc102.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * return the results list
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc102.js:19]: Great approach and high-level!

File: eliHenderson/assignments/BreadthFirstSearch/lc102.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+      if (currNode.left) queue.push(currNode.left);
+      if (currNode.right) queue.push(currNode.right);
+    }
+
+    result.push(currLvl);
+  }
+
+  return result;
+};
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc102.js:109]: Yes! Great job building that muscle memory!

File: eliHenderson/assignments/BreadthFirstSearch/lc102.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+var root = new TreeNode(12);
+root.left = new TreeNode(7);
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(9);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+console.log(`Level order traversal: ${traverse(root)}`);
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc102.js:117]: Don't forget to add test cases to match your edge cases!

File: eliHenderson/assignments/BreadthFirstSearch/lc103.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ *
+ * Time: O(N)
+ * Space: O(1)
+ *
+ *  * edge:
+ *  what if the root is null
+ *  what if the root value is null
+ *
+ */
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc103.js:41]: Fantastic approach! :100:

File: eliHenderson/assignments/BreadthFirstSearch/lc103.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      if (currentNode.left) queue.push(currentNode.left);
+      if (currentNode.right) queue.push(currentNode.right);
+    }
+
+    result.push(currLvl);
+    zigZag = !zigZag;
+  }
+
+  return result;
+};
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc103.js:122]: Muscle memory built! :100:

File: eliHenderson/assignments/BreadthFirstSearch/lc103.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+var root = new TreeNode(12);
+root.left = new TreeNode(7);
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(9);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+root.right.left.left = new TreeNode(20);
+root.right.left.right = new TreeNode(17);
+console.log(`Zigzag traversal: ${traverse(root)}`);
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc103.js:132]: Don't forget to write test cases that match your edge cases!

File: eliHenderson/assignments/BreadthFirstSearch/lc107.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ *      currlvl.push(node.val)
+ *
+ *      if(left or right child) queue. unshift( child)
+ *
+ *  result.unshift(currlvl)
+ *
+ * return result
+ *
+ * time: O(N)
+ * space: O(1)
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc107.js:36]: Great approach and nice work writing out the efficiency of this code!

File: eliHenderson/assignments/BreadthFirstSearch/lc107.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
+// Problem Statement #
+
+// Given a binary tree, populate an array to represent its level-by-level traversal in reverse order, i.e., the lowest level comes first. You should populate the values of all nodes in each level from left to right in separate sub-arrays.
+
+/**
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc107.js:5]: Missing edge cases.

File: eliHenderson/assignments/BreadthFirstSearch/lc107.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      //insert children into queue
+      if (currNode.left) queue.unshift(currNode.left);
+      if (currNode.right) queue.unshift(currNode.right);
+    }
+    //insert the currlvl into the results
+    result.unshift(currLvl);
+  }
+
+  return result;
+};
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc107.js:105]: Great job building that muscle memory!

File: eliHenderson/assignments/BreadthFirstSearch/lc107.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+var root = new TreeNode(12);
+root.left = new TreeNode(7);
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(9);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+console.log(`Reverse level order traversal: ${traverse(root)}`);
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc107.js:113]: Missing test cases that match edge cases.

File: eliHenderson/assignments/BreadthFirstSearch/lc647.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+ *  push sum / the length of the current lvl array into the results array
+ *  return result array
+ *
+ * * edge:
+ *  what if the root is null
+ *  what if the root value is null
+ *
+ * Time: O(N)
+ * Space: O(1)
+ */
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc647.js:32]: Great approach! :100:

File: eliHenderson/assignments/BreadthFirstSearch/lc647.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    //create the sum for the curr level
+    for (let j = 0; j < currLvl.length; j++) {
+      avg += currLvl[j];
+    }
+    //push the average value into the results array
+    result.push(avg / currLvl.length);
+  }
+  return result;
+};
+
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc647.js:119]: Great muscle memory workout! :100:

File: eliHenderson/assignments/BreadthFirstSearch/lc647.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+var root = new TreeNode(12);
+root.left = new TreeNode(7);
+root.right = new TreeNode(1);
+root.left.left = new TreeNode(9);
+root.left.right = new TreeNode(2);
+root.right.left = new TreeNode(10);
+root.right.right = new TreeNode(5);
+
+console.log(`Level averages are: ${find_level_averages(root)}`);
```
Comment (knrodriguez) [inline eliHenderson/assignments/BreadthFirstSearch/lc647.js:128]: Missing test cases to match edge cases.

**Review by knrodriguez — APPROVED** [review]

**Comment by knrodriguez** [comment]
Wonderful job! You are writing fantastically clear approach, calculating the time and space complexity, adding edge cases, and building your muscle memory. Just missing the test cases to match the edge cases, but once that's there it'll be perfect.
